extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java

```diff
@@ -98,220 +97,207 @@ public String fromBytes(byte[] bytes)
+            );
+
+            if (streams == null || streams.isEmpty()) {
+              throw new IAE("Topic [%s] had no streams", kafkaTopic);
             }
-          }
-      );
-      Futures.addCallback(
-          future, new FutureCallback<Object>()
-          {
-            @Override
-            public void onSuccess(Object result)
-            {
-              LOG.debug("Success listening to [%s]", topic);
+            if (streams.size() > 1) {
+              throw new ISE("Topic [%s] has %d streams! expected 1", kafkaTopic, streams.size());
             }
+            final KafkaStream<String, String> kafkaStream = streams.get(0);
 
-            @Override
-            public void onFailure(Throwable t)
-            {
-              if (t instanceof CancellationException) {
-                LOG.debug("Topic [%s] cancelled", topic);
-              } else {
-                LOG.error(t, "Error in listening to [%s]", topic);
+            // FIXME: Why is this done on every iteration?
+            startingReads.countDown();
+
+            for (final MessageAndMetadata<String, String> messageAndMetadata : kafkaStream) {
+              final String key = messageAndMetadata.key();
+              final String message = messageAndMetadata.message();
+              if (key == null || message == null) {
+                LOG.error("Bad key/message from topic [%s]: [%s]", kafkaTopic, messageAndMetadata);
+                continue;
               }
+              onNewEntry(key, message);
             }
-          },
-          MoreExecutors.sameThreadExecutor()
-      );
-      this.future = future;
-      final Stopwatch stopwatch = Stopwatch.createStarted();
-      try {
-        while (!startingReads.await(100, TimeUnit.MILLISECONDS) && connectTimeout > 0L) {
-          // Don't return until we have actually connected
-          if (future.isDone()) {
-            future.get();
-          } else {
-            if (stopwatch.elapsed(TimeUnit.MILLISECONDS) > connectTimeout) {
-              throw new TimeoutException("Failed to connect to kafka in sufficient time");
-            }
+          }
+          catch (Exception e) {
+            LOG.error(e, "Error reading stream for topic [%s]", kafkaTopic);
+          } finally {
+            Thread.interrupted();
+            consumerConnector.shutdown();
           }
         }
       }
-      catch (InterruptedException | ExecutionException | TimeoutException e) {
-        if (!future.isDone() && !future.cancel(true) && !future.isDone()) {
-          LOG.warn("Could not cancel kafka listening thread");
-        }
-        LOG.error(e, "Failed to start kafka extraction factory");
-        cacheManager.delete(factoryId);
-        return false;
-      }
+    };
+  }
 
-      started.set(true);
-      return true;
-    }
+  void onNewEntry(String key, String message) {
+    doubleEventCount.incrementAndGet();
```

Comment:
Why does it have to be double the event count? Can't it be incremented just once?